PokemonAPI/PokemonAPI.BusinessLayer/Implementations/DomainServices/BattleService.cs

```diff
@@ -0,0 +1,111 @@
+
+        public BattleService(IRepository<Pokemon> repository)
+        {
+            _repository = repository;
+        }
+
+        public async Task<BattleResult> Battle(Guid firstPokemonId, Guid secondPokemonId, CancellationToken cancellationToken = default)
+        {
+            var  firstPokemon = await _repository.Get(
+                filter: dbPokemon => dbPokemon.Id == firstPokemonId,
+                 include: source => source
+                    .Include(x => x.Abilities)
+                    .ThenInclude(x => x.Ability),
+                cancellationToken: cancellationToken
+            );
+
+            var secondPokemon = await _repository.Get(
+                filter: dbPokemon => dbPokemon.Id == secondPokemonId,
+                 include: source => source
+                    .Include(x => x.Abilities)
+                    .ThenInclude(x => x.Ability),
+                cancellationToken: cancellationToken
+            );
+
+            int GoesFirst(Pokemon firstPokemon, Pokemon secondPokemon) {
+                if (firstPokemon.BaseStats.Speed > secondPokemon.BaseStats.Speed)
+                {
+                    return 1;
+                }
+                else if (firstPokemon.BaseStats.Speed < secondPokemon.BaseStats.Speed)
+                {
+                    return 2;
+                }
+                else
+                {
+                    return 1;
+                }
+            }
+
+            int first = GoesFirst(firstPokemon, secondPokemon);
+            int firstPokeHealth = firstPokemon.BaseStats.HealthPoints;
+            int secondPokeHealth = secondPokemon.BaseStats.HealthPoints;
+            bool specialFirst = false;
+            bool specialSecond = false;
+
+            do
+            {
+                if (first == 1)
+                {
+                    firstPokeHealth = firstPokemon.BaseStats.SpecialAttack - secondPokemon.BaseStats.SpecialDefense;
+                    first = 2;
+                    specialFirst = true;
+                }
+                else
+                {
+                    secondPokeHealth = secondPokemon.BaseStats.SpecialDefense - firstPokemon.BaseStats.SpecialAttack;
+                    first = 1;
+                    specialSecond = true;
+                }
+            }
+            while (specialFirst && specialSecond && firstPokeHealth > 0 || secondPokeHealth > 0);
+
+            do
+            {
+                if (first == 1)
+                {   
+                        firstPokeHealth = firstPokemon.BaseStats.Attack - secondPokemon.BaseStats.Defense;
+                        first = 2;
+                }
+                else
+                {
+                        secondPokeHealth = secondPokemon.BaseStats.Defense - firstPokemon.BaseStats.Attack;
+                        first = 1;
+                }
+                   
+            }
+            while (firstPokeHealth > 0 || secondPokeHealth > 0);
```

Comment:
Personally in practice I prefer While loops over do while loops. As they are generally easier to implement and think about.
They also rarely fail with exceptions since you already checked things

In code block that came before

What if we have charizard vs bulbasaur. Charizard is far superior and strong and has special attack of 120. Bulbasaur is weak and small and has hp of 20.

Charizard will crush him.. bulbasaurs hp will go to - in block before this code. But you will do one pass over this DO-WHILE loop before you realize hey.. he is dead :D 

Do While works like 
Do something then repeat if while is ok.

https://dotnetfiddle.net/tfCv6W

Reply:
Also I'd recommend using datastructure called Queue which is FIFO mechanism.

First in first out.
So pokemon who goes first will be queued first and second pokemon will be queued second

Let's see what you come up with and we'll go from there